core/chains/evm/logpoller/log_poller.go

```diff
@@ -316,20 +315,20 @@ func (lp *logPoller) Replay(ctx context.Context, fromBlock int64) error {
 	}
 	// Block until replay notification accepted or cancelled.
 	select {
-	case lp.replayStart <- ReplayRequest{fromBlock, ctx}:
-	case <-lp.ctx.Done():
-		return ErrReplayAbortedOnShutdown
+	case lp.replayStart <- fromBlock:
 	case <-ctx.Done():
 		return ErrReplayAbortedByClient
 	}
 	// Block until replay complete or cancelled.
 	select {
 	case err := <-lp.replayComplete:
 		return err
-	case <-lp.ctx.Done():
-		return ErrReplayAbortedOnShutdown
 	case <-ctx.Done():
-		return ErrReplayAbortedByClient
+		// Note: this will not abort the actual replay, it just means the client gave up on waiting for it to complete
+		go func() {
+			<-lp.replayComplete
+		}()
```

Comment:
It avoids the lp.run() loop from getting stuck.  Without this, it will hang on sending to replayComplete and not be able to process any more logs after that.

Reply:
Can we use a buffered channel instead?

Reply:
Oh, I see that `replayComplete` is a long lived channel. That is actually pretty confusing. I think following standard patterns here would simplify things a lot, both in terms of lines of code and readability. e.g. the callers could pass their own `complete` channel, and then the extra struct field with broader scope wouldn't be necessary, and each one could be buffered to unblock `lp.run()`. Maybe out of scope, but seems to be complicating this PR.

Reply:
The original implementation I had in the January PR for this did something pretty similar... ironically, it seems we've come almost full circle now.   :-(  The original PR was an all-async implementation of `lp.Replay()` which was much simpler than the synchronous one we have now.  The channel was still stored in the struct, but re-initialized freshly each time lp.Replay() was called.  Then there was a separate `lp.WaitForReplayComplete()` method for synchronization.  Because the channel only got used once each time, the complete signal was just `close(lp.replayComplete)`.  So effectively, the same as a new unbuffered channel each time.

At the time, I was under the impression we wanted all calls to it in production code to be async, so this only got used in the tests.  For a while I was still convinced this was better than the context-merging approach recommended in the JIRA ticket.  But after some discussions on it, I became convinced context-merging would probably be simpler was less of a departure from the present design, involving less unknowns.  In particular, I was worried that using `lp.WaitForReplayComplete()` in production might lead to race conditions which would require additional synchronization to handle, due to the possibility of multiple threads being allowed to call each of the two methods at once.  Using just a single longterm channel seemed like a more straightforward way of ensuring things always happen in a nice sequential order... so I closed that PR and opened a separate one, which eventually grew into the one you saw.

I still think some type of short-lived channel solution might be better overall, but let's revisit later.  For now, I've implemented all of the other suggestions you have here and they seem to be working.